File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
input = sys.stdin.readline

def binary_search():
    st = 1
    en = max(tree)
    
    while st <= en:
        mid = (st + en) // 2
        cnt = 0
        for t in tree:
            if mid < t:
                cnt += (t - mid)
        if cnt >= m:
            st = mid + 1
        else:
            en = mid - 1
    return en

n,m = map(int,input().split())
tree = list(map(int,input().split()))
print(binary_search())

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] trees;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        trees = new int[N];
        for (int i = 0; i < N; i++) {
            trees[i] = Integer.parseInt(st.nextToken());
        }
        int s = 0, e = 10_0000_0000;
        while (s + 1 < e) {
            int mid = (s + e) / 2;
            if (check(mid, M)) {
                s = mid;
            } else {
                e = mid;
            }
        }
        System.out.println(s);
    }

    static boolean check(int mid, int M) {
        long sum = 0;
        for (int tree : trees) {
            if (tree > mid) {
                sum += tree - mid;
            }
        }
        return sum >= M;
    }
}